### PR TITLE
derp: add new health update and server restarting frame types

### PR DESCRIPTION
**Pull request by bradfitz**
Updates #2756
Updates #2746

### CODE REVIEW

File: derp/derp.go [hunk 1]
```diff
@@ -101,6 +101,19 @@ const (
...
+	// if they connection is unhealthy somehow. Currently the only unhealthy state
+	// is whether the connection is detected as a duplicate.
+	// The entire frame body is the text of the error message. An empty message
+	// clears the error state.
+	frameHealth = frameType(0x14)
+
+	// frameRestarting is sent from server to client for the
+	// server to declare that it's restarting. Payload is two big
+	// endian uint32 duration in milliseconds: when to reconnect,
+	// and how long to try total.
```
Comment (dsnet) [inline derp/derp.go:115]: What's the purpose of "how long to try total"? If that time period expires, what's the alternative? And if there is a valid alternative, is the client allowed to try that alternative before this period as ended?

File: derp/derp_client.go [hunk 2]
```diff
@@ -369,6 +370,38 @@ type KeepAliveMessage struct{}
...
+// connection health state.
+type HealthMessage struct {
+	// Problem, if non-empty, is a description of why the connection
+	// is unhealthy.
+	//
+	// The empty string means the connection is healthy again.
+	//
+	// The default condition is healthy, so the server doesn't
+	// broadcast a HealthMessage until a problem exists.
+	Problem string
```
Comment (crawshaw) [inline derp/derp_client.go:383]: In programming we have a thousand ways to say something is wrong, but not one way to say everything is fine.

File: derp/derp_client.go [hunk 3]
```diff
@@ -369,6 +370,38 @@ type KeepAliveMessage struct{}
...
+
+func (HealthMessage) msg() {}
+
+// ServerRestartingMessage is a one-way message from server to client,
+// advertising that the server is restarting.
+type ServerRestartingMessage struct {
+	// ReconnectIn is an advisory duration that the client should wait
+	// before attempting to reconnect. It might be zero.
+	// It exists for the server to smear out the reconnects.
+	ReconnectIn time.Duration
```
Comment (crawshaw) [inline derp/derp_client.go:394]: This is fine, but your commit message makes it sound like the client will definitely not pick a  new home. Seems reasonable that if a server says it's going away for entire seconds to pick a new home.

**Review by crawshaw — APPROVED** [review]

File: derp/derp.go (hunk 1 above)
Comment (bradfitz) [reply derp/derp.go:115]: It's explained more on the ServerRestartingMessage field docs. I'll reference those.

File: derp/derp_client.go (hunk 3 above)
Comment (crawshaw) [reply derp/derp_client.go:394]: oh I probably extrapolated from https://github.com/tailscale/tailscale/issues/2756,  which I think was in the email notification above this PR.

**Pull request merged**